sklearn/preprocessing/tests/test_data.py

```diff
@@ -851,6 +853,288 @@ def test_robust_scaler_iris_quantiles():
+                  [1, 0, 2, 2, 1]]).T
+    X_sparse = sparse.csc_matrix(X)
+    transformer = QuantileTransformer(ignore_implicit_zeros=True,
+                                      n_quantiles=5)
+
+    # dense case -> warning raise
+    assert_warns_message(UserWarning, "'ignore_implicit_zeros' takes effect"
+                         " only with sparse matrix. This parameter has no"
+                         " effect.", transformer.fit, X)
+
+    X_gt = np.array([[0, 0, 0, 0, 0],
+                     [0, 0, 1, 1, 0]]).T
+    X_trans = transformer.fit_transform(X_sparse)
+    assert_almost_equal(X_gt, X_trans.A)
+
+    # consider the case where sparse entries are missing values and user-given
+    # zeros are to be considered
+    X_data = np.array([0, 0, 1, 0, 2, 2, 1, 0, 1, 2, 0])
+    X_col = np.array([0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1])
+    X_row = np.array([0, 4, 0, 1, 2, 3, 4, 5, 6, 7, 8])
+    X_sparse = sparse.csc_matrix((X_data, (X_row, X_col)))
+    X_trans = transformer.fit_transform(X_sparse)
+    X_gt = np.array([[0., 0.5],
+                     [0., 0.],
+                     [0., 1.],
+                     [0., 1.],
+                     [0., 0.5],
+                     [0., 0.],
+                     [0., 0.5],
+                     [0., 1.],
+                     [0., 0.]])
+    assert_almost_equal(X_gt, X_trans.A)
+
+    transformer = QuantileTransformer(ignore_implicit_zeros=True,
+                                      n_quantiles=5)
+    X_data = np.array([-1, -1, 1, 0, 0, 0, 1, -1, 1])
+    X_col = np.array([0, 0, 1, 1, 1, 1, 1, 1, 1])
+    X_row = np.array([0, 4, 0, 1, 2, 3, 4, 5, 6])
+    X_sparse = sparse.csc_matrix((X_data, (X_row, X_col)))
+    X_trans = transformer.fit_transform(X_sparse)
+    X_gt = np.array([[0, 1],
+                     [0, 0.5],
+                     [0, 0.5],
+                     [0, 0.5],
+                     [0, 1],
+                     [0, 0],
+                     [0, 1]])
+    assert_almost_equal(X_gt, X_trans.A)
+    assert_almost_equal(X_sparse.A, transformer.inverse_transform(X_trans).A)
+
+
+def test_quantile_transform_dense_toy():
+    X = np.array([[0, 25, 50, 75, 100],
+                  [2, 4, 6, 8, 10],
+                  [2.6, 4.1, 2.3, 9.5, 0.1]]).T
+
+    transformer = QuantileTransformer(n_quantiles=5)
+    transformer.fit(X)
+
+    X_trans = transformer.fit_transform(X)
+    assert_almost_equal(np.min(X_trans, axis=0), 0.)
+    assert_almost_equal(np.max(X_trans, axis=0), 1.)
+
+    X_test = np.array([
+        [-1,  1,  0],
+        [101, 11, 10],
+    ])
+    expected = np.array([
+        [0, 0, 0],
+        [1, 1, 1],
+    ])
+    assert_array_almost_equal(transformer.transform(X_test), expected)
+
+    X_trans_inv = transformer.inverse_transform(X_trans)
+    assert_array_almost_equal(X, X_trans_inv)
+
+    # test subsampling
+    # FIXME: there is not comparison for the moment
+    random_state = 42
+    transformer.set_params(**{'subsample': 3,
```

Comment:
you should test subsample separately. Something that seems to work for me:
* create random array of shape (N, 1)
* sort it
* check that without subsampling, `quantile_transform` deterministically returns a linspace
* repeatedly `quantile_transform` it with subsampling while changing random state
* check that the variance under subsampling is nonzero
* check that the mean under subsampling is close (within fairly wide tolerance) to linspace